### ISSUE TITLE
Option to record for however long you want

### ISSUE THREAD

**Issue by 0K9090** [ScratchAddons/ScratchAddons#5101]
### Why this would be helpful/solve a problem

I find it weird that you can only record for a certain amount of time. Sure, you can set the seconds to 999999 but it still ends in a couple of minutes.

And i don't recall that my project stopped when it downloaded the recording

### What this change does

you can record for how ever long you want to

### Possible alternatives

download a screen recorder, but it will most likely have a watermark

**Comment by WorldLanguages** [comment]
> Sure, you can set the seconds to 999999 but it still ends in a couple of minutes.

5 minutes if the maximum. We do not properly validate the second count if it is typed, instead of using the input controls.

https://github.com/ScratchAddons/ScratchAddons/blob/master/addons/mediarecorder/userscript.js#L227

**Comment by WorldLanguages** [comment]
Do we have a public statement on why 5 min is the limit? I think we could allow users to bypass it if they know what they're doing.

**Comment by DNin01** [comment]
> Do we have a public statement on why 5 min is the limit? I think we could allow users to bypass it if they know what they're doing.

I don't know, but it might be that longer files take longer to render or process or take up more memory and that it just wouldn't be feasible after some amount of recording.

**Comment by WorldLanguages** [comment]
So I think it's mainly RAM usage (if browsers don't optimize this, all the video would be stored in RAM before it's stopped, hopefully it ends up in swap/disk) and big video file sizes.

**Comment by BroJac5246** [comment]
> So I think it's mainly RAM usage (if browsers don't optimize this, all the video would be stored in RAM before it's stopped, hopefully it ends up in swap/disk) and big video file sizes.

Could we add an option to override the limit and include a warning?

**Comment by WorldLanguages** [comment]
I think we could, but I wouldn't be sure what the best way would be.

**Comment by apple502j** [comment]
Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...

**Comment by 0K9090** [comment]
> Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...

the only solution to that would be a playstation-like-thing where if it's on for too long it will automatically turn off if the user doesn't close the prompt

**Comment by BroJac5246** [comment]
> > Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...
> 
> the only solution to that would be a playstation-like-thing where if it's on for too long it will automatically turn off if the user doesn't close the prompt

We could also check if the user is interacting with anything.

**Comment by 0K9090** [comment]
> > > Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...
> > 
> > 
> > the only solution to that would be a playstation-like-thing where if it's on for too long it will automatically turn off if the user doesn't close the prompt
> 
> We could also check if the user is interacting with anything.

what if it's one of those idle games or an animation?

**Comment by BroJac5246** [comment]
> > > > Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...
> > > 
> > > 
> > > the only solution to that would be a playstation-like-thing where if it's on for too long it will automatically turn off if the user doesn't close the prompt
> > 
> > We could also check if the user is interacting with anything.
> 
> what if it's one of those idle games or an animation?

It's just an idea, and that's what a prompt to continue recording could be used for.

**Comment by WorldLanguages** [comment]
Related: https://github.com/ScratchAddons/ScratchAddons/issues/2057#issuecomment-1249835176
We might be able to allow users to record the whole stage. The cursor, variables and lists would be visible! New API available in Chrome allows that.
I don't think this mitigates any of the potential problems of allowing recordings to last very long, however.

**Comment by WorldLanguages** [comment]
This was brought up again on the community Discord server.

**Comment by WorldLanguages** [comment]
> Do we have a public statement on why 5 min is the limit? I think we could allow users to bypass it if they know what they're doing.

Maybe we could extend it to 10 minutes, and tell the user if they've entered a number of seconds that exceeds that. Right now, if the user types 301, we silently stop after 300 seconds without warning them.

**Comment by WorldLanguages** [comment]
PR #6603 suggests increasing the limit to 10 minutes as a first measure.

**Comment by Hans5958** [comment]
> Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...

This reasoning is stupid. If that's the **only** case, why the heck we want to cap the time just because a lapse of judgement? I would be pissed that I got capped in the same way Bandicam limits its recording time on the free tier, just because "some kid will fill their disk to their brim".

We give no warranty on this software; each user is liable of their usage of the extension. If they fill their storage by their fault, it's their problem and we are not responsible to any of it.

**Comment by BroJac5246** [comment]
> > Yeah, the last thing we want is a 10GB recording because someone forgot to turn off recording...
> 
> This reasoning is stupid. If that's the **only** case, why the heck we want to cap the time just because a lapse of judgement? I would be pissed that I got capped the same way Bandicam limits its recording time on the free tier, just because "some kid will fill their disk to their brim".
> 
> We give no warranty on this software; each user is liable of their usage of the extension. If they fill their storage by their fault, it's their problem and we are not responsible to any of it.

And it's not like it takes more than 10 seconds to delete the file.

**Comment by DNin01** [comment]
There is still the RAM usage concern. Long recordings could potentially crash the page. I think it might be helpful to [get the approximate amount of memory](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/deviceMemory) with `navigator.deviceMemory`, and if it's at least 4 GB or so, there'd be an addon setting to enable longer recordings or remove the duration limit. However, that API is experimental and not supported in Firefox or Safari.

Does anyone know if we talked about the idea of labeling certain settings as dangerous or asking for confirmation before enabling such settings?

**Comment by WorldLanguages** [comment]
> Does anyone know if we talked about the idea of labeling certain settings as dangerous or asking for confirmation before enabling such settings?

That would be useful in many cases. Not sure if this is one of them, though.

**Comment by DNin01** [comment]
> > Does anyone know if we talked about the idea of labeling certain settings as dangerous or asking for confirmation before enabling such settings?
> 
> That would be useful in many cases. Not sure if this is one of them, though.

We obviously want users to be aware of the implications of having recordings go on forever.

**Comment by Samq64** [comment]
I think we should add a toggle to remove the limit but add a warning.

**Comment by DNin01** [comment]
Another way to get the amount of system memory is with `chrome.system.memory`.

The `"system.memory"` permission doesn't show any permission warnings and is much more precise. It can also give the current _available_ amount of memory, not just the capacity.

We could create an addon API that acts as a bridge so addon userscripts could access information about the device's memory capacity/free space.
